Remotion/Web/Development.Analyzers/RMWEB0003_TypeNotFromSameAssemblyWxeResourcePageStepAnalyzer.cs

```diff
@@ -0,0 +1,105 @@
+// License, or (at your option) any later version.
+//
+// re-motion is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Web.Development.Analyzers
+{
+  [DiagnosticAnalyzer(LanguageNames.CSharp)]
+  public class RMWEB0003TypeNotFromSameAssemblyWxeResourcePageStepAnalyzer : DiagnosticAnalyzer
+  {
+    private record SymbolContext (
+        INamedTypeSymbol WxeResourcePageStepSymbol,
+        INamedTypeSymbol SystemTypeSymbol);
+
+    public static readonly DiagnosticDescriptor RMWEB0003DiagnosticDescriptor = new(
+        "RMWEB0003",
+        $"Passing a type from another assembly to the constructor of WxeResourcePageStep is probably a mistake.",
+        "'{0}' is not from the same assembly as the containing type '{1}'. This might be a mistake.",
+        "Usage",
+        DiagnosticSeverity.Error,
+        true);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RMWEB0003DiagnosticDescriptor);
+
+    public override void Initialize (AnalysisContext context)
+    {
+      context.EnableConcurrentExecution();
+      context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+      context.RegisterCompilationStartAction(
+          compilationContext =>
+          {
+            var wxeResourcePageStepSymbol = compilationContext.Compilation.GetTypeByMetadataName("Remotion.Web.ExecutionEngine.WxeResourcePageStep");
+            if (wxeResourcePageStepSymbol == null)
+              return;
+
+            var systemTypeSymbol = compilationContext.Compilation.GetTypeByMetadataName("System.Type")!;
+
+            var symbolContext = new SymbolContext(
+                wxeResourcePageStepSymbol,
+                systemTypeSymbol);
+
+            compilationContext.RegisterSyntaxNodeAction(
+                ctx => AnalyzeObjectCreationExpression(ctx, symbolContext),
+                ImmutableArray.Create(SyntaxKind.ObjectCreationExpression, SyntaxKind.ImplicitObjectCreationExpression));
+          });
+    }
+
+    private void AnalyzeObjectCreationExpression (SyntaxNodeAnalysisContext ctx, SymbolContext symbolContext)
+    {
+      var node = (BaseObjectCreationExpressionSyntax)ctx.Node;
+      if (ctx.SemanticModel.GetOperation(node) is not IObjectCreationOperation operation || !SymbolEqualityComparer.Default.Equals(operation.Type, symbolContext.WxeResourcePageStepSymbol))
```

Comment:
a bit long, and since we already have so many exit conditions, splitting the or in separate if's probably won't hurt either :)